Cubical/Algebra/CommMonoid/GrothendieckGroup.agda

```diff
@@ -0,0 +1,226 @@
+
+    open GroupTheory (AbGroup→Group A)
+
+    open AbGroupStr ⦃...⦄ using (-_; _-_; _+_; invr; invl)
+    private
+      instance
+        AAsGroup  = snd A
+        MAsGroup  = snd (Groupification M)
+        AAsMonoid = snd (AbGroup→CommMonoid A)
+
+    private
+      module φ = IsMonoidHom (snd φ)
+      f = fst φ
+
+    inducedHom : AbGroupHom (Groupification M) A
+    fst inducedHom = elim (λ x → isSetAbGroup A) g proof
+        where
+        g = λ (a , b) → f a - f b
+        proof : (u v : ⟨ M² M ⟩) (r : R M u v) → g u ≡ g v
+        proof _ _ (k , p) = lemma (lemma₂ p)
+            where
+            lemma₂ : ∀ {k a b c d} → k · (a · d) ≡ k · (b · c) → f a + f d ≡ f b + f c
+            lemma₂ {k} {a} {b} {c} {d} p =
+              f a + f d   ≡⟨ sym (φ.pres· _ _) ⟩
+              f (a · d)   ≡⟨ ·CancelL _ (sym (φ.pres· _ _) ∙ (cong f p) ∙ φ.pres· _ _) ⟩
+              f (b · c)   ≡⟨ φ.pres· _ _ ⟩
+              f b + f c   ∎
+
+            lemma : ∀ {a b c d} → f a + f d ≡ f b + f c → f a - f b ≡ f c - f d
+            lemma {a} {b} {c} {d} p =
+              f a - f b
+                ≡⟨ cong (λ x → x - f b) (sym (rid _)) ⟩
+              f a + ε - f b
+                ≡⟨ cong (λ x → f a + x - f b) (sym (invr _)) ⟩
+              f a + (f d - f d) - f b
+                ≡⟨ cong (λ x → x - f b) (assoc _ _ _) ⟩
+              f a + f d - f d - f b
+                ≡⟨ cong (λ x → x - f d - f b) p ⟩
+              f b + f c - f d - f b
+                ≡⟨ comm _ _ ∙ assoc _ _ _ ⟩
+              (- f b + (f b + f c)) - f d
+                ≡⟨ cong (λ x → x - f d) (assoc _ _ _) ⟩
+              ((- f b + f b) + f c) - f d
+                ≡⟨ cong (λ x → (x + f c) - f d) (invl _) ∙ sym (assoc _ _ _) ∙ lid _ ⟩
+              f c - f d
+                ∎
+
+    pres· (snd inducedHom) = elimProp2 (λ _ _ → isSetAbGroup A _ _) proof
+      where
+        rExp : ∀ {x y z} → x ≡ y → x + z ≡ y + z
+        rExp r = cong₂ _+_ r refl
+
+        lExp : ∀ {x y z} → x ≡ y → z + x ≡ z + y
+        lExp r = cong₂ _+_ refl r
+
+        proof : ((a , b) (c , d) : ⟨ M² M ⟩) → (f (a · c)) - (f (b · d)) ≡ (f a - f b) + (f c - f d)
+        proof (a , b) (c , d) =
+          f (a · c) - f (b · d)     ≡⟨ cong₂ _-_ (φ.pres· _ _) (φ.pres· _ _) ⟩
+          (f a + f c) - (f b + f d) ≡⟨ lExp (invDistr _ _ ∙ comm _ _) ∙ assoc _ _ _ ⟩
+          ((f a + f c) - f b) - f d ≡⟨ lemma ⟩
+          (f a - f b) + (f c - f d) ∎
+          where
+            lemma = rExp (sym (assoc _ _ _) ∙ lExp (comm _ _) ∙ assoc _ _ _) ∙ sym (assoc _ _ _)
+
+    pres1 (snd inducedHom)   = invr _
+    presinv (snd inducedHom) = elimProp (λ _ → isSetAbGroup A _ _)
+                                        (λ _ → sym (invDistr _ _ ∙ cong₂ _-_ (invInv _) refl))
+
+{-
+            M ------φ------> A
+             \               Λ
+              \             /
+       universalHom    ∃! inducedHom
+                \         /
+                 V       /
+             Groupification M
+-}
+
```

Comment:
This nice diagram would be more helpful, if it were at the beginning of the module...

Reply:
Done!